copy/copy.go

```diff
@@ -580,20 +572,40 @@ func (ic *imageCopier) copyUpdatedConfigAndManifest(ctx context.Context) ([]byte
 		return nil, err
 	}
 
-	ic.c.progressPool.Wait()
 	ic.c.Printf("Writing manifest to image destination\n")
 	if err := ic.c.dest.PutManifest(ctx, manifest); err != nil {
 		return nil, errors.Wrap(err, "Error writing manifest")
 	}
 	return manifest, nil
 }
 
-// createProgressBar creates a mpb.Bar.  Note that if the copier's reportWriter
+// newProgressPool creates a *mpb.Progress and a cleanup function.
+// The caller must eventually call the returned cleanup function after the pool will no longer be updated.
+func (c *copier) newProgressPool(ctx context.Context) (*mpb.Progress, func()) {
+	ctx, cancel := context.WithCancel(ctx)
+	pool := mpb.New(mpb.WithWidth(40), mpb.WithOutput(c.progressOutput), mpb.WithContext(ctx))
+	return pool, func() {
+		cancel()
+		pool.Wait()
+	}
+}
+
+// createProgressBar creates a mpb.Bar in pool.  Note that if the copier's reportWriter
 // is ioutil.Discard, the progress bar's output will be discarded
-func (c *copier) createProgressBar(info types.BlobInfo, name, kind string) *mpb.Bar {
-	bar := c.progressPool.AddBar(info.Size,
+func (c *copier) createProgressBar(pool *mpb.Progress, info types.BlobInfo, kind string) *mpb.Bar {
+	// shortDigestLen is the length of the digest used for blobs.
+	const shortDigestLen = 12
+
+	prefix := fmt.Sprintf("Copying %s %s", kind, info.Digest.Encoded())
+	// Truncate the prefix (chopping of some part of the digest) to make all progress bars aligned in a column.
+	maxPrefixLen := len("Copying blob ") + shortDigestLen
```

Comment:
Nit: this could be made a const outside the function to avoid redundantly calculating it.

Reply:
OTOH that would move it far from the "Copying blob" string and make it easier for the two to get out of sync.

Or this could just hard-code `13` instead of `len("Copying blob ")`, but even with a comment that’s not really worth the readability impact.

This is to an extent already overthinking the previous `shortDigest()[:10]` implementation, and I’d prefer not to overthink it that much.

---

Ultimately/luckily, the compiler can already optimize `len("Copying blob ") + shortDigestLen` to `25`, so this `len` calls costs exactly zero:
```go
func prefix(digest digest.Digest, kind string) string {
	// shortDigestLen is the length of the digest used for blobs.
	const shortDigestLen = 12

	prefix := fmt.Sprintf("Copying %s %s", kind, digest.Encoded())
	// Truncate the prefix (chopping of some part of the digest) to make all progress bars aligned in a column.
	maxPrefixLen := len("Copying blob ") + shortDigestLen
	if len(prefix) > maxPrefixLen {
		prefix = prefix[:maxPrefixLen]
	}
	return prefix
}
```
(edited to remove most addresses and other garbage from the output):
```console
$ go run -gcflags '-S' foo.go
…
	CALL	fmt.Sprintf(SB)
	MOVQ	40(SP), AX  # return value of `Sprintf` == prefix, string start
	MOVQ	48(SP), CX  # return value of `Sprintf` == prefix, string length == len(prefix)
	CMPQ	CX, $25  # len(prefix) > maxPrefixLen
	JLE	291
	MOVL	$25, CX   # prefix = prefix[:maxPrefixLen]
291: 	MOVQ	AX, "".~r2+168(SP)  # returned string start
	MOVQ	CX, "".~r2+176(SP)  # returned string length
```

Reply:
Makes perfect sense :) Thanks!